### PR TITLE
midx: Introduce a parser for multi-pack-index files

### PR DESCRIPTION
**Pull request by lhchavez**
This change is the first in a series to add support for git's
multi-pack-index. This should speed up large repositories significantly.

Part of: #5399

### CODE REVIEW

File: src/multipack.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+/*
+ * An entry within the Object Offset table.
+ */
+struct git_multipack_index_object_offset {
+	/* The index of the packfile name within the Packfile Names table. */
+	uint32_t pack_int_id;
+	/* The offset in bytes within the packfile for the object. */
+	uint32_t offset;
+} __attribute__((packed));
```
Comment (lhchavez) [inline src/multipack.h]: argh, forgot this was a gcc thing. i'll remove it.

**Comment by ethomson** [comment]
Instead of strictly reverting the `sha1_lookup` change, let's leave it in `pack.c`, that way we don't have two additional files in the tree for a single function.  We can provide it internally for the library and name it `git_pack__lookup_sha1` or something.

**Comment by lhchavez** [comment]
> Instead of strictly reverting the `sha1_lookup` change, let's leave it in `pack.c`, that way we don't have two additional files in the tree for a single function. We can provide it internally for the library and name it `git_pack__lookup_sha1` or something.

good point! done! (also took the opportunity to add a comment to the function and rename the arguments).

**Comment by ethomson** [comment]
Sorry about the delay here.  Let's move forward with this.  Would you mind rebasing this on master so that we can get an up-to-date CI pass?

**Comment by lhchavez** [comment]
> Sorry about the delay here. Let's move forward with this. Would you mind rebasing this on master so that we can get an up-to-date CI pass?

Done!

**Comment by ethomson** [comment]
One thought that I have is that maybe this should be called `git_midx` instead of `git_multipack_index`.  a) `midx` is significantly shorter 😀  but b) the multipack index is the atomic unit here, if I understand correctly.  `git_multipack_index` makes it sound like there's an index on an atomic unit called a `git_multipack`.  But there is not.

Since this is an internal-only type, I don't necessarily think that it's a hard and fast requirement to make this change, but I think that the update would be an improvement.  What do you think @lhchavez?

**Comment by lhchavez** [comment]
> One thought that I have is that maybe this should be called `git_midx` instead of `git_multipack_index`. a) `midx` is significantly shorter  but b) the multipack index is the atomic unit here, if I understand correctly. `git_multipack_index` makes it sound like there's an index on an atomic unit called a `git_multipack`. But there is not.
> 
> Since this is an internal-only type, I don't necessarily think that it's a hard and fast requirement to make this change, but I think that the update would be an improvement. What do you think @lhchavez?

sounds reasonable. Done!

**Comment by ethomson** [comment]
Cooooool.  I'm excited to see this.  I'm going to chip away at your other PRs, I apologize for making you wait so long in the review queue - I haven't had much spare time in the last few months.

**Review by ethomson — APPROVED** [review]

**Pull request merged**